### PR TITLE
Consider type ordering in MergeOptimization

### LINKED ISSUE #955: Optimization error due to underspecified static shape
Still need to debug this a bit further, showed up in https://github.com/pymc-devs/pymc/pull/5757

```python
import aesara
import aesara.tensor as at
import numpy as np

w = aesara.shared(np.random.rand(20))
u = aesara.shared(np.random.rand(20))

z0 = at.arange(0, 20).astype("float64")
z0_ = z0.dimshuffle("x", 0)
hwz = at.tanh(z0_.dot(w))
z1 = z0_ + at.outer(hwz, u)
z1 = z1.flatten()
aesara.config.on_opt_error = "raise"
at.jacobian(z1, z0).eval()
```

<details>
 <summary>Traceback</summary>

```
<<!! BUG IN FGRAPH.REPLACE OR A LISTENER !!>> <class 'TypeError'> The type of the replacement (TensorType(float64, (1, None))) must be compatible with the type of the original Variable (TensorType(float64, (1, 20))). MergeOptimizer
ERROR (aesara.graph.opt): SeqOptimizer apply MergeOptimizer
ERROR (aesara.graph.opt): Traceback:
ERROR (aesara.graph.opt): Traceback (most recent call last):
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/graph/opt.py", line 280, in apply
    sub_prof = optimizer.optimize(fgraph)
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/graph/opt.py", line 103, in optimize
    ret = self.apply(fgraph, *args, **kwargs)
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/graph/opt.py", line 746, in apply
    fgraph.replace_all_validate(pairs, reason="MergeOptimizer")
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/graph/features.py", line 553, in replace_all_validate
    fgraph.replace(r, new_r, reason=reason, verbose=False, **kwargs)
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/graph/fg.py", line 511, in replace
    self.change_node_input(
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/graph/fg.py", line 428, in change_node_input
    raise TypeError(
TypeError: The type of the replacement (TensorType(float64, (1, None))) must be compatible with the type of the original Variable (TensorType(float64, (1, 20))).
Traceback (most recent call last):
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/link/vm.py", line 1112, in make_all
    node.op.make_thunk(node, storage_map, compute_map, [], impl=impl)
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/scan/op.py", line 1411, in make_thunk
    isinstance(out, TensorVariable) for out in self.fn.maker.fgraph.outputs
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/scan/op.py", line 1349, in fn
    self._fn = function(
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/compile/function/__init__.py", line 317, in function
    fn = pfunc(
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/compile/function/pfunc.py", line 363, in pfunc
    return orig_function(
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/compile/function/types.py", line 1725, in orig_function
    m = Maker(
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/compile/function/types.py", line 1464, in __init__
    optimizer_profile = optimizer(fgraph)
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/graph/opt.py", line 112, in __call__
    return self.optimize(fgraph)
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/graph/opt.py", line 103, in optimize
    ret = self.apply(fgraph, *args, **kwargs)
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/graph/opt.py", line 291, in apply
    self.failure_callback(e, self, optimizer)
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/graph/opt.py", line 226, in warn
    raise exc
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/graph/opt.py", line 280, in apply
    sub_prof = optimizer.optimize(fgraph)
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/graph/opt.py", line 103, in optimize
    ret = self.apply(fgraph, *args, **kwargs)
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/graph/opt.py", line 746, in apply
    fgraph.replace_all_validate(pairs, reason="MergeOptimizer")
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/graph/features.py", line 553, in replace_all_validate
    fgraph.replace(r, new_r, reason=reason, verbose=False, **kwargs)
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/graph/fg.py", line 511, in replace
    self.change_node_input(
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/graph/fg.py", line 428, in change_node_input
    raise TypeError(
TypeError: The type of the replacement (TensorType(float64, (1, None))) must be compatible with the type of the original Variable (TensorType(float64, (1, 20))).
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/IPython/core/interactiveshell.py", line 3418, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-37-615f4005dd2e>", line 15, in <module>
    at.jacobian(z1, z0).eval()
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/graph/basic.py", line 567, in eval
    self._fn_cache[inputs] = function(inputs, self)
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/compile/function/__init__.py", line 317, in function
    fn = pfunc(
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/compile/function/pfunc.py", line 363, in pfunc
    return orig_function(
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/compile/function/types.py", line 1736, in orig_function
    fn = m.create(defaults)
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/compile/function/types.py", line 1631, in create
    _fn, _i, _o = self.linker.make_thunk(
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/link/basic.py", line 254, in make_thunk
    return self.make_all(
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/link/vm.py", line 1121, in make_all
    raise_with_op(fgraph, node)
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/link/utils.py", line 538, in raise_with_op
    raise exc_value.with_traceback(exc_trace)
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/link/vm.py", line 1112, in make_all
    node.op.make_thunk(node, storage_map, compute_map, [], impl=impl)
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/scan/op.py", line 1411, in make_thunk
    isinstance(out, TensorVariable) for out in self.fn.maker.fgraph.outputs
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/scan/op.py", line 1349, in fn
    self._fn = function(
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/compile/function/__init__.py", line 317, in function
    fn = pfunc(
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/compile/function/pfunc.py", line 363, in pfunc
    return orig_function(
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/compile/function/types.py", line 1725, in orig_function
    m = Maker(
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/compile/function/types.py", line 1464, in __init__
    optimizer_profile = optimizer(fgraph)
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/graph/opt.py", line 112, in __call__
    return self.optimize(fgraph)
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/graph/opt.py", line 103, in optimize
    ret = self.apply(fgraph, *args, **kwargs)
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/graph/opt.py", line 291, in apply
    self.failure_callback(e, self, optimizer)
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/graph/opt.py", line 226, in warn
    raise exc
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/graph/opt.py", line 280, in apply
    sub_prof = optimizer.optimize(fgraph)
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/graph/opt.py", line 103, in optimize
    ret = self.apply(fgraph, *args, **kwargs)
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/graph/opt.py", line 746, in apply
    fgraph.replace_all_validate(pairs, reason="MergeOptimizer")
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/graph/features.py", line 553, in replace_all_validate
    fgraph.replace(r, new_r, reason=reason, verbose=False, **kwargs)
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/graph/fg.py", line 511, in replace
    self.change_node_input(
  File "/home/ricardo/Documents/Projects/pymc3-venv/lib/python3.8/site-packages/aesara/graph/fg.py", line 428, in change_node_input
    raise TypeError(
TypeError: The type of the replacement (TensorType(float64, (1, None))) must be compatible with the type of the original Variable (TensorType(float64, (1, 20))).
Apply node that caused the error: for{cpu,scan_fn}(TensorConstant{20}, TensorConstant{[ 0  1  2 .. 17 18 19]}, TensorConstant{20}, Reshape{1}.0, InplaceDimShuffle{x,0}.0, Elemwise{Composite{(i0 - sqr(i1))}}[(0, 1)].0, InplaceDimShuffle{0,x}.0)
Toposort index: 8
Inputs types: [TensorType(int64, ()), TensorType(int64, (20,)), TensorType(int64, ()), TensorType(float64, (None,)), TensorType(float64, (1, None)), TensorType(float64, (1,)), TensorType(float64, (None, 1))]
HINT: Use a linker other than the C linker to print the inputs' shapes and strides.
HINT: Re-running with most Aesara optimizations disabled could provide a back-trace showing when this node was created. This can be done by setting the Aesara flag 'optimizer=fast_compile'. If that does not work, Aesara optimizations can be disabled with 'optimizer=None'.
HINT: Use the Aesara flag `exception_verbosity=high` for a debug print-out and storage map footprint of this Apply node.
```

</details>

### PR DESCRIPTION
**Pull request by ricardoV94**
Closes #955

### CODE REVIEW

**Comment by ricardoV94** [comment]
This is actually overkill, the problem is that `type.convert_variable` it this line is too weak and can only change the broadcastable info: 

https://github.com/aesara-devs/aesara/blob/35bf0d353f9681341fe320d2847df87520afc4d8/aesara/graph/opt.py#L773

This will no longer be an issue after #915 gets merged

File: aesara/graph/opt.py [hunk 1]
```diff
@@ -638,6 +638,38 @@ def process_node(self, fgraph, node):
                     # They were already tried, and there was an error
                     continue
 
+                # Check if outputs of node are all super types of the candidate outputs
+                if not all(
+                    node_out.type.is_super(candidate_out.type)
+                    for node_out, candidate_out in zip(node.outputs, candidate.outputs)
```
Comment (brandonwillard) [inline aesara/graph/opt.py:644]: Given how often merging is done, we should be careful about performance in these areas, so let's see if we can reformulate this condition and the following one to involve only a single pass through the candidates.

**Review by brandonwillard — COMMENTED** [review]
This looks good; just a small performance-related concern to address.

File: aesara/graph/opt.py (hunk 1 above)
Comment (ricardoV94) [reply aesara/graph/opt.py:644]: I didn't spot a/the place where we can control the order over which we process nodes. In any  case I am not too happy with this and would rather just let #915 take care of it.

Comment (brandonwillard) [reply aesara/graph/opt.py:644]: Couldn't both directions be computed in one pass over `zip(node.outputs, candidate.outputs)`?

For example:
```python
                type_orders_1, type_orders_2 = zip(
                    *(
                        (
                            node_out.type.is_super(candidate_out.type),
                            candidate_out.type.is_super(node_out.type),
                        )
                        for node_out, candidate_out in zip(
                            node.outputs, candidate.outputs
                        )
                    )
                )
                # Check if outputs of node are all super types of the candidate outputs
                if not all(type_orders_1):
                    # If not, check if reverse is true
                    if not all(type_orders_2):
                        continue
```

Comment (ricardoV94) [reply aesara/graph/opt.py:644]: Ahh I thought you mean't the recalling process_node logic. 

I actually don't know how often we have this situation of mergeable nodes with different output types. I assumed it was rather low since it only showed in one very specific PyMC test so far, so that's why I kept the checks separately

Also this way it short circuits with `all`

Comment (brandonwillard) [reply aesara/graph/opt.py:644]: First off, I don't know if those two distinct `all` conditions are even necessary.  It's possible that one of the `is_super` calls can be short-circuited (e.g. using `or` between the two) in a single pass through the pairs that also subsumes the two `if` conditions.

Comment (ricardoV94) [reply aesara/graph/opt.py:644]: I need to know which of the two is True, one is to proceed as usual, the other is to reverse things. In any case I am much more worried about the performance penalty of recalling `process_node` than the second loop of type comparisons.

Comment (ricardoV94) [reply aesara/graph/opt.py:644]: A quick benchmark suggests the zip option to be slower:

```python
import aesara.tensor as at

def separate_loop(vars1, vars2):
    type_orders1 = (
        var1.type.is_super(var2.type)
        for var1, var2 in zip(vars1, vars2)
    )
    type_orders2 = (
        var2.type.is_super(var1.type)
        for var2, var1 in zip(vars2, vars1)
    )
    return all(type_orders1) or all(type_orders2)

def symultaneous_loop(vars1, vars2):
    type_orders_1, type_orders_2 = zip(
        *(
            (
                var1.type.is_super(var2.type),
                var2.type.is_super(var1.type),
            )
            for var1, var2 in zip(vars1, vars2)
        )
    )

    return all(type_orders_1) or all(type_orders_2)


type1 = at.TensorType("float64", shape=(1, 2, None, 3))
type2 = at.TensorType("float64", shape=(1, 2, None, None))
vars1 = (type1(), type1(), type1())
vars2 = (type2(), type2(), type2())

%timeit separate_loop(vars1, vars2)
%timeit separate_loop(vars2, vars1)
%timeit symultaneous_loop(vars1, vars2)
%timeit symultaneous_loop(vars2, vars1)
4.59 µs ± 31.9 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
3.51 µs ± 6.79 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
6.42 µs ± 130 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
6.36 µs ± 32.7 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

Comment (ricardoV94) [reply aesara/graph/opt.py:644]: I don't see how the zip case can ever be faster from those results. Also I don't quite get what is the performance concern with the second separate loop.

Comment (ricardoV94) [reply aesara/graph/opt.py:644]: I did try different numbers and orders this was actually one of the more comparable cases

Comment (brandonwillard) [reply aesara/graph/opt.py:644]: > I don't see how the zip case can ever be faster from those results. Also I don't quite get what is the performance concern with the second separate loop.

The `zip` _isn't_ faster, and it wasn't supposed to be; it was meant to illustrate that both conditions can be handled with a single pass.

The approach I'm talking about simplifies the two conditions using a single pass over the pairs.  The example I gave doesn't actually do that, though.  It only illustrates how both conditions are related via a single pass.  The next step involves some consideration of what `is_super` and those two conditions are attempting to do.

Comment (brandonwillard) [reply aesara/graph/opt.py:644]: Recall that `TensorType.is_super` is supposed to be a partial ordering.  More specifically, it serves the purpose of `>=` between types, which is something like a product order.  (I believe the docstring for `Type.is_super` says `>`, so those are out of sync.)

With that in mind, it looks like your first condition checks that not all `a_n >= b_n` (i.e. that at least one instance of `a_n < b_n` is true) across all `n`, and the second condition checks that not all `b_n >= a_n` are true (i.e. that at least one instance of `b_n < a_n` is true).

In this setting, the general assumption is that the `Type`s are generally equal except for differences in their static shape specificity, and we can reduce the problem to a few simple cases:
1. all of the `candidate` outputs' static shapes are equal or more specific than `node`'s,
2. all of the `candidate` outputs' static shapes are less specific than `node`'s,
3. or some mix of the two.

The implicit `else` branch of the first `if` condition appears to handle 1. (i.e. queue the replacement), the implicit `else` branch of the second `if` handles 2. (i.e. attempt to queue the reverse replacement), and the first branch of the second `if` handles 3. (i.e. ignore the replacement).

The formulation of these three cases makes it seem like a second `is_super` pass is necessary, as well as the more costly recursive calls to `process_node`; however, the basic idea is that we want to replace less specific `Type`s with more specific&mdash;yet equivalent&mdash;ones (i.e. when `a <= b` use `Type` `a` instead of `b`).  

Without these case constraints we could apply the `<=`/`is_super` ordering across each pair of corresponding outputs (only _once_) and pick the "smallest"/most specific `Type` in each case.  If we could do this, then we would also avoid discarding candidates at this stage (which we never did previously) and the aforementioned performance concerns.

The apparent challenge here is that the most informative set of output `Type`s for the corresponding `Op` and inputs might not already exist as an `Apply` node in the graph (i.e. it isn't `node` or `candidate`); however, nothing is preventing us from creating and using such an `Apply` node.

In other words, if we track signatures comprised of `Op`, inputs, and outputs (with the "smallest" `Type`s) tuples&mdash;instead of effectively just the first two&mdash;we could consistently narrow the `Type`s in a graph and only perform a single `min`-based pass over node outputs and avoid recursion.

**Comment by ricardoV94** [comment]
Closing this, as https://github.com/aesara-devs/aesara/pull/915 solves the original bug. This PR would provide a more efficient graph (with less SpecifyShape's), but add extra complexity in the MergeOptimizer rewrite that may not be warranted.
